fish-rust/src/wildcard.rs

```diff
@@ -11,3 +48,1249 @@ pub const ANY_STRING_RECURSIVE: char = char_offset(WILDCARD_RESERVED_BASE, 2);
+                    self.expand_intermediate_segment(
+                        base_dir,
+                        &mut dir,
+                        wc_segment,
+                        wc_remainder,
+                        &(effective_prefix.to_owned() + wc_segment + L!("/")),
+                    );
+                } else {
+                    // Last wildcard segment, nonempty wildcard.
+                    self.expand_last_segment(base_dir, &mut dir, wc_segment, effective_prefix);
+                }
+
+                let Some(asr_idx) = wc_segment.find_char(ANY_STRING_RECURSIVE) else {
+                    return;
+                };
+
+                // Apply the recursive **.
+                // Construct a "head + any" wildcard for matching stuff in this directory, and an
+                // "any + tail" wildcard for matching stuff in subdirectories. Note that the
+                // ANY_STRING_RECURSIVE character is present in both the head and the tail.
+                let head_any = wc_segment.slice_to(asr_idx + 1);
+                let any_tail = wc.slice_from(asr_idx);
+                assert!(head_any.chars().last().unwrap() == ANY_STRING_RECURSIVE);
+                assert!(any_tail.chars().next().unwrap() == ANY_STRING_RECURSIVE);
+
+                dir.rewind();
+                self.expand_intermediate_segment(
+                    base_dir,
+                    &mut dir,
+                    head_any,
+                    any_tail,
+                    effective_prefix,
+                );
+            }
+        }
+
+        pub fn status_code(&self) -> WildcardResult {
+            if self.did_interrupt {
+                return WildcardResult::Cancel;
+            } else if self.did_overflow {
+                return WildcardResult::Overflow;
+            } else if self.did_add {
+                WildcardResult::Match
+            } else {
+                WildcardResult::NoMatch
+            }
+        }
+    }
+
+    impl<'e> WildCardExpander<'e> {
+        /// We are a trailing slash - expand at the end.
+        fn expand_trailing_slash(&mut self, base_dir: &wstr, prefix: &wstr) {
+            if self.interrupted_or_overflowed() {
+                return;
+            }
+
+            if !self.flags.contains(ExpandFlags::FOR_COMPLETIONS) {
+                // Trailing slash and not accepting incomplete, e.g. `echo /xyz/`. Insert this file, we already know it exists!
+                self.add_expansion_result(base_dir.to_owned());
+                return;
+            }
+            // Trailing slashes and accepting incomplete, e.g. `echo /xyz/<tab>`. Everything is added.
+            let Ok(mut dir) = self.open_dir(base_dir, false) else {
+                return;
+            };
+
+            // wreaddir_resolving without the out argument is just wreaddir.
+            // So we can use the information in case we need it.
+            let need_dir = self.flags.contains(ExpandFlags::DIRECTORIES_ONLY);
```

Comment:
missing this part:

        if (flags & expand_flag::special_for_cd) {
            path = this->working_directory;
            append_path_component(path, base_dir);
            // cd operates on logical paths.
            // for example, cd ../<tab> should complete "without resolving symlinks".
            path = normalize_path(path);
        }

looks like we don't have a test for it

Reply:
nevermind there's already a thread